GitCommands/Git/GitModule.cs

```diff
@@ -1208,9 +1212,19 @@ bool IsTwoSha1Hashes(string parents)
             }
         }
 
+        [CanBeNull]
         public ConfigFile GetSubmoduleConfigFile()
         {
-            return new ConfigFile(WorkingDir + ".gitmodules", true);
+            try
+            {
+                return new ConfigFile(WorkingDir + ".gitmodules", true);
+            }
+            catch (Exception ex)
+            {
+                ExceptionUtils.ShowException(ex, false);
```

Comment:
Multiple issues here.

1. This is a very poor UX, a user won't know what to do with this.
Nor the team would be in a position to provide a meaningful help.

    ![image](https://user-images.githubusercontent.com/14078392/67165787-26ca3080-f357-11e9-85cc-236138981bde.png)
    
    The error must be concise and clear. Something like:
    ```
    Failed to parse .gitmodules config
    -------------------------------------
    
    An error occurred while parsing <path>\.gitmodules
    
    Please fix the config and reopen the repository.
    ```

2. UI rendering must only be done at the UI layer (i.e. Form or a Control).
The backend must raise exceptions in a way that is useful to callers, e.g. throw a specific exception.

    We can add a new exception (e.g. `GitConfigurationException`) to contain necessary information:
    ```cs
    public class GitConfigurationException : Exception
    {
        public GitConfigurationException(string configPath, Exception inner) : base(null, inner)
        {
            if (string.IsNullOrWhiteSpace(configPath))
            {
                throw new ArgumentException("ConfigPath is required", nameof(configPath));
            }

            ConfigPath = configPath;
        }

        public string ConfigPath { get; }
    }
    ```

    and call sites should handle the exception as necessary.

Reply:
@RussKie 
I am having a little bit of trouble getting this new exception class to propagate to the UI layer.  The UI layer seems to be making calls to the SubmoduleStatusProvider, which makes an async call to GitCommands.Submodules.SubmoduleStatusProvider.UpdateSubmodulesStatusAsync.  I am unable to catch exceptions from async code on the UI thread, and am not sure how to proceed.

Reply:
@aarnott / @sharwell I don't suppose there is a way to intercept/route an exception from a `FileAndForget` call to the main thread other than the following?

```cs
void MyMethod(Action<Exception> exceptionHandler)
{
    ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
    {
        try 
        {
			await TaskScheduler.Default;
			RunCommand();
		}
		catch (Exception)
		{
			await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancelToken);
			exceptionHandler(ex);
		}
    }).FileAndForget();
}
```

Reply:
I'll first respond to the @russkie's suggestion of the simpler message. While I support a simpler message box to present to the user, the simpler message @russkie proposed *lacks* the detail that would help the user to fix the problem. Specifically the message about the invalid escape character would be *very* useful in helping the user fix the issue. What I'll usually do to craft useful error messages is for the top-level catch block that's responsible to turn the exception into a message box to concatenate the exception messages from the outer exception into the deepest inner exception, from outer to inner, as a complete sentence each. This usually produces a very user-friendly error message like "Error initializing git submodules. Error while parsing the .gitmodule file. Failure to parse a \d escape sequence."

Reply:
Now to the question about how to handle exceptions on the main thread. Yes for a 'fire and forget' method (which is usually something I try to avoid anyway, but is sometimes useful) it can catch its own exception and switch to the main thread before popping a message box just as you've done it. At least in your sample, it isn't clear from the method that the `Action<Exception>` delegate has thread affinity or which thread it would like to run on. So switching to the main thread before invoking it seems a bit odd. If you have xml doc comments on the method explaining this behavior and callers rely on that, then that's fine. Otherwise I'd consider making it a `Func<Exception, Task>` delegate so that the handler can conveniently switch to the thread it needs (if at all).

Reply:
Thank you. 
1. Totally agree, if an exception's hierarchy contains useful to the end user information - it should be shown.
2. The last snippet was pretty much written on a napkin. Indeed switching to the UI thread should occur in the exception handler.

Reply:
@lhiginbotham so I suppose we have the following plan

1. Create a specific exception, that is thrown if unable to parse the .gitmodules files
2. In sync execution flows - callsites should handle the exception at UI layer (i.e. at `GiUI` project and not in `GitCommands`)
```cs
try 
{
    // execute functionality that parses .gitmodules file
}
catch (GitConfigurationException ex)
{
    string message = ... // build a good error message
    MessageBox.Show(....);
}
```
3. In async flows, like the above, call sites must pass a delegate that would execute in the UI layer. The delegate will need to switch to the UI thread 
```cs
// callsite
Func<Exception, Task> showErrorAsync = async (exception) =>
{
    await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancelToken);

    string message = ... // build a good error message
    MessageBox.Show(....);
};
_submoduleStatusProvider.UpdateSubmodulesStatus(Module.WorkingDir, status, showErrorAsync);

// worker
        public void UpdateSubmodulesStatus(string workingDirectory, [CanBeNull] IReadOnlyList<GitItemStatus> gitStatus, Func<Exception, Task> showErrorAsync)
        {
            //...

            var cancelToken = _submodulesStatusSequence.Next();
            _gitStatusWhileUpdatingStructure = null;
            ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
            {
                await TaskScheduler.Default;

                var currentModule = new GitModule(workingDirectory);

                try
                {
                    await UpdateSubmodulesStatusAsync(currentModule, gitStatus, cancelToken);
                }
                catch (Exception ex)
                {
                    await showErrorAsync(ex);
                }

                OnStatusUpdated(_submoduleInfoResult, cancelToken);
            }).FileAndForget();
        }
```

What do you think?

NB: writing in a browser, some things may be off, hopefully the general idea stands.

Reply:
That looks better. But why define a fire-and-forget method that takes a callback? That seems like you're writing old-style (pre-async) code. Why not just return `Task` from your method and let the caller await it with a try/catch around it?

Reply:
@RussKie 
I'll give your latest code suggestion a try - one thing that I have noticed so far is that I need to update all of the unit tests that use this newly modified UpdateSubmodulesStatus interface, which brings two concerns on my part:

1. Should the unit tests provide a meaningful showErrorAsync callback to display any legit errors, or should it just 'do nothing'?

2. How can I switch to the main thread in the showErrorAsync's provided by the unit tests?  I first thought that I needed to create a cancellation token (was going to try `CancellationTokenSource cts = new CancellationTokenSource();` and use `cts` to get a cancellation token) but it turns out that in the following pattern used in the main GitExtensions code, `ThreadHelper` is a class defined in the `GitUI` namespace:

    `await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(token);`

    I can try to simply add `using GitUI;` to the top of the unit test file (SubmoduleStatusProviderTests.cs), but I've noticed that this test seems to strictly be related to GitCommands - it felt wrong getting GitUI involved.  I guess all of this stems from my poor familiarity with async code/practices and really boils down to whether it is necessary for the showErrorAsync to switch to the main thread in the unit test's case if we are just going to swallow the exception.

Reply:
> Why not just return `Task` from your method and let the caller await it with a try/catch around it?

I suppose we can do that, thank you.
@gerhardol do you see any issues?

Reply:
> 2\. How can I switch to the main thread in the showErrorAsync's provided by the unit tests?

You don't, in unit tests you don't show any UI, so there is no need to be a main thread.
Unit test just need to assert a handler is called with expected data.

If we change from a `FileAndForget` method, there won't be any handler to pass, the method will just throw.
Let's continue the discussion.

Reply:
I would not change all the calls to `FileAndForget`: The caller wants to trigger an action whose result/success is not of great interest (as say function name says). I'd rather handle the exceptions inside `FileAndForget`: There could be a generic, common pipe to the main thread in order to display exceptions.

Reply:
I see no issue in changing this. The current tests are not testing any useful anyway.
But a popup from low level should be OK here, this is misformatted data anyway.

Reply:
@RussKie 

While attempting to take care of all synchronous calls that depend on the .gitsubmodules file parser, I've come to the realization that the config file parser is nearly at the entire root of GitExtensions.  There are so many dependencies on this config file parser that this entire PR might just turn into "Fix Issue #WXYZ: represent _all_ exceptions with a message box from the UI layer".  Do you think I should attempt to put try-catchs all throughout the code-base, or should I swallow some of these errors to cut down on some dependency-trees whenever possible? (For example, in methods that produce `IEnumerable`s based off of the result of parsing the git submodules config file, return an empty list instead of throwing an exception)

Reply:
What do you have in mind?

We do have handlers for unhandled exceptions, those are hooked to NBug form. So we would generally know when things go pear shape. I don't think it would be wise to put plumbing for handling all possible exceptions, because most of them probably won't happen in real life, or users won't be able to fix them themselves.

In this instance we want to handle a specific exception which users observed, and that _can_ happen in real life. And we want to give users guidance on how they can fix the error themselves.

IIUIC, there are few sync flows, where it would be a matter of handling the exception and showing a good message. In one instance, we need to change a `FileAndForget` method into an `async` and handle the exception as well.
Am I missing anything?

Reply:
>> There could be a generic, common pipe to the main thread in order to display exceptions.

> We do have handlers for unhandled exceptions, those are hooked to NBug form.

Yes, of course (was quite late yesterday).

Exceptions like `GitConfigurationException` should be shown to the user with a meaningful error message. I think a NBug report is not necessary in this case, especially GE should not be terminated if the error is not fatal.

If there were many code paths which could throw such "user error message" exceptions without a need for special handling, there could be inserted a common handler, which displays these exceptions, before the NBug handler.

> In one instance, we need to change a `FileAndForget` method into an `async` and handle the exception as well.

Yes, that's better than let `FileAndForget` return `Task`.